proxy/proxy.go

```diff
@@ -177,149 +161,8 @@ func (s *ProxyServer) Start() {
-	if r.ContentLength > s.config.Proxy.LimitBodySize {
-		log.Printf("Socket flood from %s", ip)
-		s.policy.ApplyMalformedPolicy(ip)
-		http.Error(w, "Request too large", http.StatusExpectationFailed)
-		return
-	}
-	r.Body = http.MaxBytesReader(w, r.Body, s.config.Proxy.LimitBodySize)
-	defer r.Body.Close()
-
-	cs := &Session{ip: ip, enc: json.NewEncoder(w)}
-	dec := json.NewDecoder(r.Body)
-	for {
-		var req JSONRpcReq
-		if err := dec.Decode(&req); err == io.EOF {
-			break
-		} else if err != nil {
-			log.Printf("Malformed request from %v: %v", ip, err)
-			s.policy.ApplyMalformedPolicy(ip)
-			return
-		}
-		cs.handleMessage(s, r, &req)
-	}
-}
-
-func (cs *Session) handleMessage(s *ProxyServer, r *http.Request, req *JSONRpcReq) {
-	if req.Id == nil {
-		log.Printf("Missing RPC id from %s", cs.ip)
-		s.policy.ApplyMalformedPolicy(cs.ip)
-		return
-	}
-
-	vars := mux.Vars(r)
-	login := strings.ToLower(vars["login"])
-
-	if !util.IsValidHexAddress(login) {
-		errReply := &ErrorReply{Code: -1, Message: "Invalid login"}
-		cs.sendError(req.Id, errReply)
-		return
-	}
-	if !s.policy.ApplyLoginPolicy(login, cs.ip) {
-		errReply := &ErrorReply{Code: -1, Message: "You are blacklisted"}
-		cs.sendError(req.Id, errReply)
-		return
-	}
-
-	// Handle RPC methods
-	switch req.Method {
-	case "eth_getWork":
-		reply, errReply := s.handleGetWorkRPC(cs)
-		if errReply != nil {
-			cs.sendError(req.Id, errReply)
-			break
-		}
-		cs.sendResult(req.Id, &reply)
-	case "eth_submitWork":
-		if req.Params != nil {
-			var params []string
-			err := json.Unmarshal(req.Params, &params)
-			if err != nil {
-				log.Printf("Unable to parse params from %v", cs.ip)
-				s.policy.ApplyMalformedPolicy(cs.ip)
-				break
-			}
-			reply, errReply := s.handleSubmitRPC(cs, login, vars["id"], params)
-			if errReply != nil {
-				cs.sendError(req.Id, errReply)
-				break
-			}
-			cs.sendResult(req.Id, &reply)
-		} else {
-			s.policy.ApplyMalformedPolicy(cs.ip)
-			errReply := &ErrorReply{Code: -1, Message: "Malformed request"}
-			cs.sendError(req.Id, errReply)
-		}
-	case "eth_getBlockByNumber":
-		reply := s.handleGetBlockByNumberRPC()
-		cs.sendResult(req.Id, reply)
-	case "eth_submitHashrate":
-		cs.sendResult(req.Id, true)
-	default:
-		errReply := s.handleUnknownRPC(cs, req.Method)
-		cs.sendError(req.Id, errReply)
-	}
-}
-
-func (cs *Session) sendResult(id json.RawMessage, result interface{}) error {
-	message := JSONRpcResp{Id: id, Version: "2.0", Error: nil, Result: result}
-	return cs.enc.Encode(&message)
-}
-
-func (cs *Session) sendError(id json.RawMessage, reply *ErrorReply) error {
-	message := JSONRpcResp{Id: id, Version: "2.0", Error: reply}
-	return cs.enc.Encode(&message)
-}
-
-func (s *ProxyServer) writeError(w http.ResponseWriter, status int, msg string) {
-	w.WriteHeader(status)
-	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+func (s *ProxyServer) rpc(level int) *rpc.RPCClient {
+	return s.upstreams[level]
```

Comment:
Maybe enhance this to make it bounds-safe?

Reply:
As long as there is no way for someone externally to invoke this and cause a crash, then I say leave it as is. If we misuse this function, it will cause a crash during development and we can fix our usage of it.

Just don't want it to be triggerable from a remote machine

Reply:
That makes sense, wouldn't want this to mask any issues.